reporter/src/main/kotlin/reporters/OpossumReporter.kt

```diff
@@ -0,0 +1,605 @@
+    val isDir = isDirectory == true || (isDirectory == null && (if (right != "") right else left).last() == '/')
+    return "/$left/$right"
+        .replace(Regex("/*$"), if (isDir) "/" else "")
+        .replace(Regex("/+"), "/")
+        .ifEmpty { "/" }
+}
+
+fun resolvePath(pieces: List<String>): String {
+    return pieces.reduce { right, left -> resolvePath(right, left) }
+}
+
+/**
+ * A [Reporter] that generates an [Opossum Input].
+ *
+ * This reporter supports the following options:
+ * - *scanner.maxDepth*: The depth to which the full file level scanner information is added
+ * - *scopes.excluded*: Comma seperated list of scopes that are excluded
+ */
+class OpossumReporter : Reporter {
+    companion object {
+        const val OPTION_SCANNER_MAXDEPTH = "scanner.maxDepth"
+        const val OPTION_EXCLUDED_SCOPES = "scopes.excluded"
+    }
+
+    data class OpossumSignal(
+        val source: String,
+        val id: Identifier? = null,
+        val url: String? = null,
+        val license: SpdxExpression? = null,
+        val copyright: String? = null,
+        val comment: String? = null,
+        val preselected: Boolean = false,
+        val followUp: Boolean = false,
+        val excludeFromNotice: Boolean = false,
+        val uuid: UUID = UUID.randomUUID()
+    ) {
+        fun toJson(): Map<*, *> {
+            return sortedMapOf(
+                uuid.toString() to sortedMapOf(
+                    "source" to sortedMapOf(
+                        "name" to source,
+                        "documentConfidence" to 80
+                    ),
+
+                    "attributionConfidence" to 80,
+
+                    "packageType" to id?.getPurlType(),
+                    "packageNamespace" to id?.namespace,
+                    "packageName" to id?.name,
+                    "packageVersion" to id?.version,
+
+                    "copyright" to copyright,
+                    "licenseName" to license?.toString(),
+
+                    "url" to url,
+
+                    "preSelected" to preselected,
+                    "followUp" to if (followUp) "FOLLOW_UP" else null,
+                    "excludeFromNotice" to excludeFromNotice,
+
+                    "comment" to comment
+                )
+            )
+        }
+
+        fun matches(other: OpossumSignal): Boolean =
+            source == other.source
+                    && id == other.id
+                    && url == other.url
+                    && license == other.license
+                    && copyright == other.copyright
+                    && comment == other.comment
+                    && preselected == other.preselected
+    }
+
+    data class OpossumResources(
+        val tree: MutableMap<String, OpossumResources> = mutableMapOf()
+    ) {
+        fun addResource(pathPieces: List<String>) {
+            if (pathPieces.isEmpty()) {
+                return
+            }
```

Comment:
Nit: We prefer to format short single-expression conditionals as one-liners. Similar in several places below.